src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +32,131 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Applies changes to the given document so that the folders, name, and path from <paramref name="newDocumentInfo"/>
+        /// are used. Will also perform symbol changes to the code in the document, attempting to sync type and namespace
+        /// to the new information.
+        /// 
+        /// Note that regardless of symbol changes, the document info will be applied and returned in the new solution. 
+        /// </summary>
+        /// <remarks>
+        /// Casing changes in document info will be applied but will not result in any symbol changes.
+        /// </remarks>
+        /// <returns><see cref="RenameDocumentResult"/> with the modified solution and indication of any symbol changes made</returns>
+        public static async Task<RenameDocumentResult> RenameDocumentAsync(
+            Document document, DocumentInfo newDocumentInfo, OptionSet optionSet, CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentInfo.FilePath == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentInfo.FilePath));
```

Comment:
This needs a message, or otherwise the exception is just going to say "FilePath" and that won't look good.